distributed/tests/test_nanny.py

```diff
@@ -956,3 +956,33 @@ async def test_nanny_plugin_register_nanny_killed(c, s, restart):
     finally:
         proc.kill()
     assert await register == {}
+
+
+@pytest.mark.slow
+@gen_cluster(
+    client=True,
+    Worker=Nanny,
+    nthreads=[("", 1)],
+    worker_kwargs={"heartbeat_interval": "10ms"},
+)
+async def test_nanny_does_not_restart_worker_on_graceful_retirement(c, s, a):
+    """Tests https://github.com/dask/distributed/pull/8522
+
+    Some clusters (e.g. SpecCluster) implement downscaling by calling
+    `Scheduler.retire_workers()` without arguments, which defaults to
+     `remove=True, close_workers=False`.
+
+    and then use an external system to tear down the worker and the nanny. In these
+    cases, make sure that the worker doesn't kill itself and that the nanny doesn't
+    restart it after the heartbeat to the scheduler fails.
+    """
+    await s.retire_workers([a.worker_address], stimulus_id="test")
+    # On Linux, it takes ~3.5s for the nanny to resuscitate a worker
+    await asyncio.sleep(5)
```

Comment:
This will either be missed entirely on github actions or will make the test very flaky. If you are waiting for some condition to occur, please wait until that condition did occur. Just having a plain sleep in here is not sufficient. Besides, this makes it also much harder to understand the test logic.

Reply:
> This will either be missed entirely on github actions 

Yes, getting a false negative on GH actions is a possibility.

>  or will make the test very flaky.

No, because if GH is slower than my local machine it will simply not test the use case properly and return false negative.

> If you are waiting for some condition to occur,

No, I am waiting for a condition NOT to occur.

I will rewrite the unit test to check for the heartbeat stop. It will no longer verify that heartbeat _does not_ call close().